homeassistant/components/zwave_js/discovery_data_template.py

```diff
@@ -422,3 +427,123 @@ def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]
+
+    def __post_init__(self) -> None:
+        """Validate inputs."""
+        assert len(self.speeds) > 0
+
+
+@dataclass
+class FixedFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate, FixedFanSpeedValueMix
+):
+    """
+    Specifies a fixed set of fan speeds.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=FixedFanSpeedDataTemplate(
+              speeds=[32,65,99]
+          ),
+      ),
+
+    `speeds` indicates the maximum setting on the underlying fan controller
+    for each actual speed.
+    """
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int]:
+        """Get the fan speed configuration for this device."""
+        return self.speeds
+
+
+@dataclass
+class ConfigurableFanSpeedValueMix:
+    """Mixin data class for defining configurable fan speeds."""
+
+    configuration_option: ZwaveValueID
+    configuration_value_to_speeds: dict[int, list[int]]
+
+    def __post_init__(self) -> None:
+        """Validate inputs."""
+        for speeds in self.configuration_value_to_speeds.values():
+            assert len(speeds) > 0
+
+
+@dataclass
+class ConfigurableFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate, ConfigurableFanSpeedValueMix
+):
+    """
+    Gets fan speeds based on a configuration value.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=ConfigurableFanSpeedDataTemplate(
+            configuration_option=ZwaveValueID(
+                5, CommandClass.CONFIGURATION, endpoint=0
+            ),
+            configuration_value_to_speeds={0: [32, 65, 99], 1: [24, 49, 74, 99]},
+          ),
+      ),
+
+    `configuration_option` is a reference to the setting that determines how
+    many speeds are supported.
+
+    `configuration_value_to_speeds` maps the values from `configuration_option`
+    to a list of speeds.  The specified speeds indicate the maximum setting on
+    the underlying switch for each actual speed.
+    """
+
+    def resolve_data(self, value: ZwaveValue) -> dict[str, ZwaveConfigurationValue]:
+        """Resolve helper class data for a discovered value."""
+        zwave_value = self._get_value_from_id(value.node, self.configuration_option)
+        return {"configuration_value": zwave_value}
+
+    def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]:
+        """Return list of all ZwaveValues that should be watched."""
+        return [
+            resolved_data["configuration_value"],
+        ]
+
+    def get_speed_config(
+        self, resolved_data: dict[str, ZwaveConfigurationValue]
+    ) -> list[int]:
+        """Get current speed configuration from resolved data."""
+        zwave_value: ZwaveValue = resolved_data["configuration_value"]
+
+        if zwave_value.value is None:
```

Comment:
If the Value.value is None it may make sense to make the entity unavailable since the entity features depend on this value. We shouldn't raise uncaught exceptions here intentionally since it would leak to the entity state update and we don't have a nice user experience for that. This is not a developer error but a something is unexpected with the device state and we need to handle it but also inform the user.

Reply:
K, done.